### PR TITLE
Adding option to shuffle shards before splitting to workers, based on current epoch

### PR DESCRIPTION
**Pull request by gabrielilharco**
### What
This PR adds an option to shuffle the shards **before** splitting them between nodes and workers. We coordinate this shuffling via a shared random seed, to ensure shards are properly distributed between workers. With epoch-based training loops, the idea is to set this random seed to a different number every epoch, ensuring shards are distributed differently at each iteration. This is completely optional, and should not affect existing workflows in any way. If users chose to use it, the only change they need to make is to simply call `dataloader.set_epoch(epoch)` at each epoch before iterating.

### Why
Without this, each worker always sees the same subset of shards at every epoch, when using the standard `nodesplitter` and `split_by_worker`. We found this to have a drastic negative impact in performance when training contrastive models, since it limits the possibilities of which datapoints are being contrasted. With this fix, this issue was substantially mitigated. 

### How
The fix is two small modifications. One is to introduce a `set_epoch` method, which informs the dataloader (and everything that composes it up to the ShardList) of the current epoch. The second one is to use that as a random seed for shuffling shards *before* they are being split between workers and nodes. 

We have found this to significantly improve performance in our contrastive learning experiments, and would greatly appreciate if you could take a look at this!

cc @mitchnw

### CODE REVIEW

**Comment by tmbdev** [comment]
Thanks; let me have a look at it. I'm not sure that this really needs a new method on WebDataset. Why don't you simply make the `nodesplitter` function dependent on the epoch itself and add a `set_epoch` function to that?

I'll see whether I can document this better and add examples.

(Also note that these issues are specific to actually performing splitting by nodes; if you use sampling with replacement, i.e. all shards on all nodes, none of this arises. That's how we usually train.)

**Comment by gabrielilharco** [comment]
Thank you for the response!

A new `nodesplitter` would also do the trick here. This way we don't need to change the logic inside `ShardList` or `Composable`. The downside of this solution is that users would need to interact with another object in their training loops (since they would need to keep track of the splitter in addition to the dataloader itself). The upside is minimal interference with existing classes. If this is preferred, we could add something like the following to `workerenv.py`:

```python
class RandomNodeSplitter:
   """Randomly split shards between nodes.
  
   Used as a shard selection function in Dataset.
   In epoch-based training loops, this splitter can be used to ensure shards are distributed
   differently among nodes at each epoch. A typical use is to call `set_seed(epoch)` before
   iterating over the dataset in each epoch.
   """
   def __init__(self, start_seed=0):
      self.set_seed(start_seed)
    
   def set_seed(self, seed):
      self.seed = seed
   
   def __call__(self, urls, env=None):
      random.Random(self.seed).shuffle(list(urls))
      return split_by_node(urls, env)
```

While this problem indeed only arises in specific use cases, splitting by nodes in epoch-based training loops is not an uncommon scenario. A good example is contrastive learning, which has gained a lot of popularity recently. In this paradigm, it is still common to have epoch-based training loops since sampling with replacement can be problematic, leading to pairing the same instance as a negative sample of itself. I agree that this fix could be done outside the library, but this seems like a common enough use case that it could benefit the library to have native support for it, especially if its done with minimal interference to existing workflows.

Let me know your thoughts!

**Comment by tmbdev** [comment]
I'll just add the `RandomNodeSplitter` class for now, which should point people in the right direction. I'll also add an example to the documentation.

A couple of comments:
- I think the PR doesn't work, because a WebDataset is a pipeline of `IterableDataset` instances, not an single object
- there are shorthands for common multinode configurations (`multimode=`) now, and I may eventually add such a shorthand for this case as well.

**Comment by tmbdev** [comment]
Roughly, the way PyTorch handles spawn is that the workers are spawned in subprocesses where torch.distributed is not initialized; in addition, the Dataset instances are restored from pickled data on every epoch. That also means that `set_epoch` in your patch wouldn't work as expected.  The usual arguments to DataLoader (shuffle, sampler, batch_sampler) are explicitly disabled in DataLoader for IterableDataset, so they don't help us.

I have refactored the code now in a way in which this should work more like you expect and that does also allow you to use per-epoch shuffling.

If you don't need per-epoch shuffling, just don't do anything.

If you do want `epoch_shuffle`, you can create the dataset instance as follows:

```
    shardlist = functools.partial(wds.PytorchShardList, epoch_shuffle=True)
    dataset = wds.WebDataset("imagenet-{000000..000015}.tgz", shardlist=shardlist). ...
    loader = wds.WebLoader(dataset, num_workers=4, batch_size=20)
```

And then you must put the epoch in the environment:

```
    loader = make_loader()
    for epoch in range(2):
        os.environ["WDS_EPOCH"] = str(epoch)
        for inputs, labels in loader:
            ...
```

You can find a preliminary example notebook here 

https://github.com/webdataset/webdataset/blob/master/notebooks/multinode-test.ipynb

**Comment by gabrielilharco** [comment]
Thank you so much for the response and updating the code!!

Your refactoring seems to also do what we need, i'll go ahead and test it out. Is there any reason why the default of `shuffle` of `PytorchShardList` is set to `False`? I also noticed the code crashes if we set it to `True` because `self.epoch` is not defined (see [https://github.com/webdataset/webdataset/blob/master/webdataset/dataset.py#L244](https://github.com/webdataset/webdataset/blob/master/webdataset/dataset.py#L244))

**Comment by tmbdev** [comment]
Thanks for spotting that. I'm not sure how the self.epoch got past the test case. I think I have fixed it. I'll try to add some test cases for it.

**Comment by gabrielilharco** [comment]
Thanks for the responsiveness here! I tested the new code and it matches the performance of the previous solutions we suggested. I'll go ahead and close this PR.

Thanks again for your help!
